### PR TITLE
Safe RAFT cluster recovery in response to nodes changing their addresses during startup

### PR DESCRIPTION
**Pull request by redouan-rhazouani**
### What's being changed:

- This PR enhances the robustness and safety of the Raft cluster recovery process, particularly when node IP addresses change.  Raft now attemps to resolve the server id before resorting to its stored addresses.
- Additionally, it enhance the boostrapping mechanism to handle ip changes during bootstrapping phase
- Replaces the current un-safe implementation, which relies on peers.json to restore the cluster

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: cloud/store/bootstrap.go [hunk 1]
```diff
@@ -106,6 +112,17 @@ func (b *Bootstrapper) notify(ctx context.Context, servers []string) (err error)
 	return
 }
 
+// servers retrieves a list of server addresses based on their names and ports.
+func (b *Bootstrapper) servers(candidates []string, serverPortMap map[string]int) []string {
+	candidates = candidates[:0]
```
Comment (moogacs) [inline cloud/store/bootstrap.go:117]: **suggestion:** may be we use `clear(candidates)` basically it's the same but could improve readability

**Review by moogacs — COMMENTED** [review]
Nice implementation!, just lift minor suggestions and clarifications

File: adapters/handlers/rest/configure_api.go [hunk 2]
```diff
@@ -243,28 +243,39 @@ func MakeAppState(ctx context.Context, options *swag.CommandLineOptionsGroup) *s
...
-		IsLocalHost: appState.ServerConfig.Config.Cluster.Localhost,
+		WorkDir:            filepath.Join(appState.ServerConfig.Config.Persistence.DataPath, "raft"),
+		NodeID:             nodeName,
+		Host:               addrs[0],
+		RaftPort:           appState.ServerConfig.Config.Raft.Port,
+		RPCPort:            appState.ServerConfig.Config.Raft.InternalRPCPort,
+		ServerName2PortMap: server2port,
+		BootstrapExpect:    appState.ServerConfig.Config.Raft.BootstrapExpect,
+		HeartbeatTimeout:   appState.ServerConfig.Config.Raft.HeartbeatTimeout,
+		RecoveryTimeout:    appState.ServerConfig.Config.Raft.RecoveryTimeout,
```
[the remark this answers is not in the document]
Comment (redouan-rhazouani) [reply adapters/handlers/rest/configure_api.go:268]: Thanks for catching this, I will remove it next time

File: cloud/store/transport.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	"time"
+
+	"github.com/hashicorp/raft"
+)
+
+// addressResolver resolves server id into an ip
+type addressResolver interface {
+	// NodeAddress resolves node id into an ip address without the port
+	NodeAddress(id string) string
+}
```
[the remark this answers is not in the document]
Comment (redouan-rhazouani) [reply cloud/store/transport.go:26]: I consider this a dependency, so using an interface here seems more appropriate. Additionally, it provides flexibility for adding more methods in the future as needed

File: cloud/store/transport.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	maxPool int, timeout time.Duration,
+) (*raft.NetworkTransport, error) {
+	cfg := &raft.NetworkTransportConfig{
+		ServerAddressProvider: r,
+		MaxPool:               tcpMaxPool,
+		Timeout:               tcpTimeout,
+	}
+
+	return raft.NewTCPTransportWithConfig(bindAddr, advertise, cfg)
+}
```
[the remark this answers is not in the document]
Comment (redouan-rhazouani) [reply cloud/store/transport.go:73]: I prefer having related functionalities together for improved readability. In this case, the store delegates this responsibility to its helper.

File: cloud/store/transport.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	return &addrResolver{
+		addressResolver:  cfg.AddrResolver,
+		RaftPort:         cfg.RaftPort,
+		IsLocalCluster:   cfg.IsLocalHost,
+		NodeName2PortMap: cfg.ServerName2PortMap,
+	}
+}
+
+// ServerAddr resolves server ID to a RAFT address
+func (a *addrResolver) ServerAddr(id raft.ServerID) (raft.ServerAddress, error) {
```
[the remark this answers is not in the document]
Comment (redouan-rhazouani) [reply cloud/store/transport.go:46]: While we could mention that it satisfies this interface, I consider it to be a private code that makes complete sense when considering its use in `NewTCPTransport`

File: cloud/store/transport.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	bindAddr string, advertise net.Addr,
+	maxPool int, timeout time.Duration,
+) (*raft.NetworkTransport, error) {
+	cfg := &raft.NetworkTransportConfig{
+		ServerAddressProvider: r,
+		MaxPool:               tcpMaxPool,
+		Timeout:               tcpTimeout,
+	}
+
+	return raft.NewTCPTransportWithConfig(bindAddr, advertise, cfg)
```
[the remark this answers is not in the document]
Comment (redouan-rhazouani) [reply cloud/store/transport.go:72]: Thanks

File: cloud/store/bootstrap.go (hunk 1 above)
Comment (redouan-rhazouani) [reply cloud/store/bootstrap.go:117]: I favor this approach over creating a new function solely for returning a slice. 
The `clear(candidate)` function does not actually clear anything; it simply returns a new slice. Therefore, slicing is makes more sense to me

**Review by moogacs — APPROVED** [review]

File: cloud/store/bootstrap.go (hunk 1 above)
Comment (moogacs) [reply cloud/store/bootstrap.go:117]: i meant `clear` the builtin function

**Pull request merged**